### DISCUSSION TITLE
How to pass *varargs?

### DISCUSSION THREAD

**Discussion by henon** [pythonnet/pythonnet#1772] [Q&A] [answered]
How do you pass a variable list of arguments in pythonnet?  The problem is the function `numpy.gradient(f, *varargs, axis=None, edge_order=1)` as documented here: https://numpy.org/doc/stable/reference/generated/numpy.gradient.html

Here is an example in python

```python
>>> import numpy as np
>>> dx=4.0
>>> dy=5.0
>>> zX=[[1,2,3],[4,5,6],[8,9,0]]
>>> np.gradient(zX, dx, dy)
[array([[ 0.75 ,  0.75 ,  0.75 ],
       [ 0.875,  0.875, -0.375],
       [ 1.   ,  1.   , -1.5  ]]), array([[ 0.2,  0.2,  0.2],
       [ 0.2,  0.2,  0.2],
       [ 0.2, -0.8, -1.8]])]
>>>
```

I tried to pass the variable args with the args tuple but then the python function complains that the number of arguments is incorrect. 

```text
Python.Runtime.PythonException: TypeError : invalid number of arguments
 ['  File "C:\\Users\\henon\\AppData\\Local\\python-3.7.3-embed-amd64\\lib\\numpy\\lib\\function_base.py", line 1013, in gradient\n    raise TypeError("invalid number of arguments")\n']   at Python.Runtime.PyObject.Invoke(PyTuple args, PyDict kw)
   at Python.Runtime.PyObject.InvokeMethod(String name, PyTuple args, PyDict kw)
```

I also tried to pass it as kw["varargs"] but then python says there is no kwarg named "varargs"

So how to call that function?

**Comment by henon** [comment]
I have an update, looks like I can pass a second tuple after the args tuple and then I am getting a different error so this is a step in the right direction. But I am still stumped as to what the function really expects 

```csharp
       using (Py.GIL()) {
                dynamic np = Py.Import("numpy");
                dynamic zX = np.array(new[,] { { 1, 2, 3 }, { 4, 5, 6 }, { 8, 9, 0 } });
                Console.WriteLine(zX.__repr__());
               var result=(np as PyObject).InvokeMethod("gradient", 
                       new PyTuple(new PyObject[] { zX as PyObject}), 
                       new PyTuple(new PyObject[] { new PyFloat(4.0), new PyFloat(5.0), })
              );
            Console.WriteLine(result.InvokeMethod("__repr__"));
    }
```

Exception:
```
Python.Runtime.PythonException: ValueError : when 1d, distances must match the length of the corresponding dimension
 ['  File "C:\\Users\\henon\\AppData\\Local\\python-3.7.3-embed-amd64\\lib\\numpy\\lib\\function_base.py", line 1004, in gradient\n    raise ValueError("when 1d, distances must match "\n']   at Python.Runtime.PyObject.Invoke(PyObject[] args)
   at Python.Runtime.PyObject.InvokeMethod(String name, PyObject[] args)
```

Any ideas @filmor ?

**Reply by henon** [reply]
Nah, looks like the second tuple is interpreted as kwargs because I can't pass additional kw args:

```csharp
                var result=(np as PyObject).InvokeMethod("gradient", 
                        new PyTuple(new PyObject[] { zX as PyObject}), 
                        new PyTuple(new PyObject[] { new PyFloat(4.0), new PyFloat(5.0), }),
                        new PyDict() { ["edge_order"]=new PyInt(1) }
                );
```
This again leads to invalid arguments error `Python.Runtime.PythonException: TypeError : invalid number of arguments`

**Comment by lostmsu** [comment]
This should work according to my understanding:
```csharp
InvokeMethod('gradient', new PyTuple(f, varArg1, varArg2), kwArgs)
```

**Reply by henon** [reply]
Yeah, that was what I would have expected too but it does not. That was what I tried initially. It gives `TypeError : invalid number of arguments` 
I am getting the feeling this is a bug.

**Comment by filmor** [comment, marked as answer]
You can see the available overloads here: https://github.com/pythonnet/pythonnet/blob/090ff9f460016b006f49c328115cc47f80be3b1f/src/runtime/PythonTypes/PyObject.cs#L815-L930

So we have

* `PyTuple`
* `params PyObject[]`
* `PyObject[], PyDict`
* `PyTuple, PyDict`

It's not ideal as it only has overloads with a `PyObject` name for the first two functions and the .NET varargs will probably hide the single `PyTuple` overload (not sure about that).

I'm not sure why you'd think any of your attempts here should work. The variant given by @lostmsu could maybe also run into selecting the wrong overload (the `params` one), I'd have to check the rules.

Does any of these work?

* `np.gradient(zX, 4.0, 5.0)` (using the fact that `np` is `dynamic`, without additional `kwArgs`)
* `np.InvokeMethod(new PyObject[] {zX, new PyFloat(4.0), new PyFloat(5.0)}, kwArgs)`

**Reply by henon** [reply]
Thanks for your time looking into this. Both your suggestions don't work. The dynamic version gives `Python.Runtime.PythonException: TypeError : invalid number of arguments` as does the second option. Do you think pythonnet is missing an overload for this kind of python function?

**Reply by henon** [reply]
I reopened this as an issue, as it is impossible to pass the *varargs. #1804

**Reply by henon** [reply]
OK, I was wrong, @filmor 's answer works with the current preview version of pythonnet, it just had a typo. The working code is:

```csharp
var kwArgs = new PyDict() { ["edge_order"] = new PyInt(1) };
var result = (np as PyObject).InvokeMethod("gradient", new PyObject[] { (zX as PyObject), new PyFloat(4.0), new PyFloat(5.0) }, kwArgs );
```
